src/compiler/debug.rs

```diff
@@ -7,6 +7,11 @@ use crate::classic::clvm::__type_compatibility__::{sha256, Bytes, BytesFromType}
 use crate::compiler::sexp::SExp;
 use crate::util::u8_from_number;
 
+/// Given an SExp and a transformation, make a map of the transformed subtrees of
+/// the given SExp in code that's indexed by treehash.  This will merge equivalent
+/// subtrees but the uses to which it's put will generally work well enough.
+///
+/// Given how it's used downstream, there'd be no way to disambiguate anyhow.
```

Comment:
```suggestion
/// Given an SExp and a transformation, make a map of the transformed subtrees of
/// the given SExp, indexed by treehash.  Note that this will merge equivalent subtrees.
///
/// Identical CLVM subtrees cannot be distinguished, assuming identical environments.
```

Reply:
This is purely syntactic so there's no environment in play here, basically just about the CLVM value space and how program source code is represented in CLVM values.

These are all equivalent in CLVM:

    "Y" Y 89 0x59

So a user writing:

    (list Y "Y" 89 0x59) ;; 1

Gives the compiler back a CLVM expression that could mean any of these things:

    (c Y (c Y (c Y (c Y ()))))
    (c "Y" (c "Y" (c "Y" (c "Y" ()))))
    (c 89 (c 89 (c 89 (c 89 ()))))
    (c 0x59 (c 0x59 (c 0x59 (c 0x59 ()))))

So the compiler rehydrates this result by taking the largest matching subtrees from the user's input and replacing it.  The above is a pathological case for this, and in general, doing something like:

   (if (some-condition X) (do-something-a X) (let ((Y (something X))) (do-something-else Y)))

Expands into a macro invocation for if, and comes back with 3 subtrees identical to the user's input, so those whole trees return with their source locations and the form of the user's input (Ys not rewritten as the number 89, but as identifiers).

It's not particularly worse than any other apporach and I always planned to make this facility better, as below.

This branch (hopefully to PR today) [https://github.com/Chia-Network/clvm_tools_rs/tree/20230227-defmac](defmac) introduces a macro form that uses the evaluator, which allows those values to pass through properly and be distinguishable in macro space.

Reply:
Great explanation. I'd say, just put your whole explanation above in the comment!